File: OCDemo/OCDemo/设计模式/CustomView.m
//
//  CustomView.m
//  OCDemo
//
//  Created by mkrq-yh on 2019/3/20.
//  Copyright © 2019 mkrq-yh. All rights reserved.
//

#import "CustomView.h"

@implementation CustomView

- (instancetype)initWithFrame:(CGRect)frame withVC:(UIViewController *)vc
{
    self = [super initWithFrame:frame];
    if (self) {
        [self setupSubViews];
    }
    return self;
}

- (void)setupSubViews {
    [_bgView removeFromSuperview];
    _bgView = [[UIView alloc] initWithFrame:self.bounds];
    [self addSubview:_bgView];
    
    UIButton *btn2 = [UIButton buttonWithType:UIButtonTypeCustom];
    btn2.frame = CGRectMake(20, 20, 200, 50);
    [btn2 setTitle:@"发送通知" forState:UIControlStateNormal];
    [btn2 setTitleColor:[UIColor blackColor] forState:UIControlStateNormal];
    btn2.titleLabel.font = [UIFont systemFontOfSize:16];
    btn2.backgroundColor = [UIColor greenColor];
    btn2.tag = 101;
    [btn2 addTarget:self action:@selector(btnLClick:) forControlEvents:UIControlEventTouchUpInside];
    [_bgView addSubview:btn2];
    
}

- (void)btnLClick:(UIButton *)sender {
    if (self.delegate && [self.delegate respondsToSelector:@selector(selectedItemButton:)]) {
        [self.delegate selectedItemButton:sender.tag];
    }
}

/*
// Only override drawRect: if you perform custom drawing.
// An empty implementation adversely affects performance during animation.
- (void)drawRect:(CGRect)rect {
    // Drawing code
}
*/

@end

File: SwiftDemo/SwiftDemo/设计模式/CustomView.swift
//
//  CustomView.swift
//  SwiftDemo
//
//  Created by mkrq-yh on 2019/3/20.
//  Copyright © 2019 mkrq-yh. All rights reserved.
//

import UIKit

protocol CustomViewDelegate {
    func selectedItemButton(index:Int)
}

class CustomView: UIView {

    var delegate: CustomViewDelegate?
    let bgView:UIView
    
    override init(frame: CGRect) {
        bgView = UIView(frame: frame)
        super.init(frame: frame)
        self.setupSubViews()
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    func setupSubViews() {
        bgView.frame = self.bounds
        self.addSubview(bgView)
        
        let btn = UIButton(frame: CGRect(x: 10, y: 10, width: 100, height: 50))
        btn.backgroundColor = UIColor.green
        btn.tag = 101
        btn.setTitle("点击", for: .normal)
        btn.setTitleColor(UIColor.red, for: .normal)
        btn.titleLabel?.font = UIFont.systemFont(ofSize: 16)
        btn.addTarget(self, action: #selector(btnClick(sender:)), for: .touchUpInside)
        bgView.addSubview(btn)
    }
    
    @objc func btnClick(sender:UIButton) {
        delegate?.selectedItemButton(index: sender.tag)
    }
    /*
    // Only override draw() if you perform custom drawing.
    // An empty implementation adversely affects performance during animation.
    override func draw(_ rect: CGRect) {
        // Drawing code
    }
    */

}
